### PR TITLE
Add support for index_select OP (frontend implementation)

### PR DESCRIPTION
**Pull request by azazhu**
This PR is frontend implementation of index_select(lookup_tv, dim, index_tv) fusion: 
- support arbitrarily rank
- support arbitrarily dim value
- index_tv can become consumer;
- lookup_tv cannot become consumer.

This PR depends on index_select frontend PR https://github.com/csarofeen/pytorch/pull/2183

OP def is here: https://pytorch.org/docs/stable/generated/torch.index_select.html

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 1]
```diff
@@ -1514,6 +1514,36 @@ class IrParser {
       }
     }
 
+    {
+      auto ptr_op = getOperatorForLiteral(
+          "aten::index_select(Tensor self, int dim, Tensor index) -> Tensor");
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: I think enablement in TorchScript by default is a big change because it means direct exposure to the user. I don't think we have done sufficient validation to do so. However, we can add an environment flag in https://github.com/csarofeen/pytorch/blob/c9f8c1de5f3bd6ac0a0cce50569410abe2dfc209/torch/csrc/jit/codegen/cuda/utils.h#L89
So that it is enabled only when a user manually does so. This way, we can both continue our development with this flag on, and start using it in PyG by manually enabling it, and at the same time avoid expose to all users by default.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 2]
```diff
@@ -1513,6 +1513,42 @@ class IrParser {
...
+            {
+              MemoryFormat format;
+              std::list<Val*> list_val;
+              std::tie(format, list_val) = getPWFormatValues(
+                  c10::nullopt,
+                  value_map[node->inputs()[0]->unique()],
+                  value_map[node->inputs()[2]->unique()]);
+              auto input = list_val.front();
+              list_val.pop_front();
+              auto dim_value = constant_as<int>(node->input(1));
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:1531]: if `int dim` is supposed to be constant, it would be nice to add it as a profiled-constant.

e.g. https://github.com/pytorch/pytorch/blob/ee05f47bddfb97b4b292808543d928b3526fc0ca/torch/csrc/jit/codegen/cuda/parser.cpp#L4233-L4251

**Review by jjsjann123 — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 2 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:1531]: Let's also add tests for this here:
https://github.com/pytorch/pytorch/blob/master/test/test_jit_cuda_fuser.py

something simple as this vvv
https://github.com/pytorch/pytorch/blob/master/test/test_jit_cuda_fuser.py#L4140-L4153

We should also test for 1). `dim` being a constant vs 2). `dim` is a runtime input.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:1531]: I think we also need an entry here for type inference: https://github.com/pytorch/pytorch/blob/ee05f47bddfb97b4b292808543d928b3526fc0ca/torch/csrc/jit/codegen/cuda/type_inference.cpp#L457

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 3]
```diff
@@ -62,6 +62,7 @@ bool isSimpleMap(Node* node) {
       "aten::logit(Tensor self, float? eps=None) -> Tensor",
       "aten::lerp(Tensor self, Tensor end, Scalar weight) -> Tensor",
       "aten::lerp(Tensor self, Tensor end, Tensor weight) -> Tensor",
+      "aten::index_select(Tensor self, int dim, Tensor index) -> Tensor",
```
Comment (jjsjann123) [inline torch/csrc/jit/passes/graph_fuser.cpp]: Nope you don't want it here. This is legacy fuser not nvfuser.

File: torch/csrc/jit/passes/guard_elimination.cpp [hunk 4]
```diff
@@ -337,6 +337,7 @@ struct GuardElimination {
       case aten::log10:
       case aten::frac:
       case aten::lerp:
+      case aten::index_select:
```
Comment (jjsjann123) [inline torch/csrc/jit/passes/guard_elimination.cpp]: why do we need to have index_select here to remove the guard?

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 5]
```diff
@@ -1513,6 +1513,42 @@ class IrParser {
...
+              TORCH_INTERNAL_ASSERT(
+                  dim_value.has_value(), "dim in index_select is not valid");
+              auto index = list_val.front();
+              list_val.pop_front();
+              Val* out = nullptr;
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: nitpick: Can we just put an `assert(false)` here with a `not implemented`? I suspect this will give us a segfault that's going to be confusing if index_select is somehow accidentally enabled.

File: torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h [hunk 6]
```diff
@@ -1266,6 +1267,41 @@ struct ReductionOpRecord : RecordFunctor {
...
+    return new IndexSelectOpRecord(*this);
+  }
+
+  void operator()(FusionDefinition& fd) final {
+    auto arg1 =
+        fd.getFusionState(args_.at(0).index)->template as<Nvf::TensorView>();
+    auto arg3 =
+        fd.getFusionState(args_.at(1).index)->template as<Nvf::TensorView>();
+
+    Nvf::Val* output = nullptr;
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h]: ditto.

**Review by jjsjann123 — COMMENTED** [review]
Looks about right.

There seems to be some unrelated changes under `torch/csrc/jit/passes/`, let's revert those.

File: test/test_jit_cuda_fuser.py [hunk 7]
```diff
@@ -4152,6 +4152,50 @@ def t(x):
...
+    def test_index_select_runtime_dim(self):
+        lookup_size = 68
+        feat_dim = 128
+        num_elements = 355984
+        dim = torch.tensor(0, device='cuda').to(dtype=torch.int)
+        lookup_tv = torch.rand(lookup_size, feat_dim, dtype=torch.float, device="cuda")
+        indies_tv = torch.randint(0, lookup_size, (num_elements,), dtype=torch.float, device="cuda").to(dtype=torch.int)
+        sbf = torch.rand(num_elements, feat_dim, dtype=torch.float, device="cuda")
+
+        def failure(x_kj, idx_kj, sbf):
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py]: We should be able to support runtime scalar input. e.g. https://github.com/pytorch/pytorch/blob/391b593ca262432ccba1939f7448275cfd4f62e6/test/test_jit_cuda_fuser.py#L1106-L1110 

runtime_dim shouldn't be a failure if the scalar is passed properly.
```
def func(x_kj, idx_kj, sbf, dim: int):
    sbf_res = torch.index_select(x_kj, dim, idx_kj) * sbf
    sbf_res = sbf_res + 17
    return sbf_res
```

And you would also need to run it like `self._run_helper(t_jit, t, lookup_tv, indies_tv, sbf, 0)`

**Review by jjsjann123 — APPROVED** [review]
LGTM, minor comments on proper test for runtime scalar, looks like the test is still not quite right, but code logic looks about right. I'm stamping it.
